src/support/small_set.h

```diff
@@ -248,7 +248,7 @@ template<typename T, size_t N, typename FlexibleSet> class SmallSetBase {
                      ConstIterator,
                      typename FlexibleSet::const_iterator>(parent) {}
 
-    const value_type operator*() const {
+    const value_type& operator*() const {
```

Comment:
Hmm, I can't make this compile... I get this:
```
src/support/small_set.h:236:16: error: binding reference of type 'wasm::SmallSetBase<wasm::LocalSet *, 2, std::set<wasm::LocalSet *, std::less<wasm::LocalSet *>, std::allocator<wasm::LocalSet *>>>::value_type' (aka 'wasm::LocalSet *') to value of type 'wasm::LocalSet *const' drops 'const' qualifier
        return *this->flexibleIterator;
               ^~~~~~~~~~~~~~~~~~~~~~~
src/ir/LocalGraph.cpp:261:16: note: in instantiation of member function 'wasm::SmallSetBase<wasm::LocalSet *, 2, std::set<wasm::LocalSet *, std::less<wasm::LocalSet *>, std::allocator<wasm::LocalSet *>>>::Iterator::operator*' requested here
  auto* aSet = *aSets.begin();
               ^
1 error generated.
```
There is no `const` anywhere in the relevant code. The erroring line has no const on it, not does the non-constant iterator which is used here... I have no idea what const it is trying to cast from or to here...

Reply:
Do you need to redefine `value_type` to `const T` or something similar to that?

Reply:
Even in the `const_iterator`, `value_type` stays the same; `reference` and `pointer` change. But yeah I don't understand what this happens either... I'm not saying this should block landing or anything but just curious what's wrong here. Why does this try to bind the type to `wasm::LocalSet *const`? 😨

Reply:
@tlively Did you mean to add that change to the const or non-const iterator? Adding it to the non-const actually does make it compile, which surprises me... is that what we want, to make the non-const iterator iterate on const values? 

(Adding it to the const iterator does not fix compilation.)

Reply:
Yes, actually the non-const iterator should iterate on const values because you should not be able to mutate values inside the set.

Reply:
I guess I don't understand C++ then... isn't the point of a non-const iterator to potentially modify the values? We do use that pattern in binaryen, I'm pretty sure - is it wrong?
```
for (auto& value : values) {
  if (..) value = foo;
}
```

Reply:
For example of it that I found:

https://github.com/WebAssembly/binaryen/blob/a34746c6708cd3e9d0c6ae554cd6bd40d43bb5b8/src/passes/Inlining.cpp#L949

+

https://github.com/WebAssembly/binaryen/blob/a34746c6708cd3e9d0c6ae554cd6bd40d43bb5b8/src/passes/Inlining.cpp#L974

Reply:
I'm confused too... Isn't the diff between `iterator` and `const_iterator` the ability to modify the elements?

Reply:
@kripken By the way I'm confused why this comment thread is attached here; the current code compiles. What doesn't compile is the `iterator` counterpart of this I think? I mean, this, at line 239:
```
    value_type& operator*() const { // With & this doesn't compile
      if (this->usingFixed) {
        return (this->parent->fixed)[this->fixedIndex];
      } else {
        return *this->flexibleIterator;
      }
    }
```

Reply:
@aheejin Maybe I don't understand your question, but yes, the current code compiles - it is adding a `&` for the non-const iterator's `operator*` that would break compilation. (The const iterator has a `&` for its `operator*`, and that compiles ok.)

Reply:
In general `iterator` modifies elements and `const_iterator` cannot modify elements, but specific collections can have stronger rules. For example `std::set` and `std::unordered_set` require `iterator` to be [constant](https://en.cppreference.com/w/cpp/container/set/begin#:~:text=Because%20both%20iterator%20and%20const_iterator%20are%20constant%20iterators%20(and%20may%20in%20fact%20be%20the%20same%20type)%2C%20it%20is%20not%20possible%20to%20mutate%20the%20elements%20of%20the%20container%20through%20an%20iterator%20returned%20by%20any%20of%20these%20member%20functions.) as well because modifying the values in place would break the set invariants (i.e. the ordering in the red-black tree or the index of the has bucket, respectively). Similarly, `iterator` on `std::map` and `std::unordered_map` allows modifying the values of the entries, but not the keys.

Reply:
Makes sense, thanks! Then we can make the return type of `Iterator::operator*` `const value_type&` too, which makes the need for two separate classes `Iterator` and `ConstIterator` unnecessary I think?

Reply:
Oh, thanks! Yes I guess that makes sense, changing the value would indeed be bad... TIL

Fixed, using a const iterator in both places, and now it all seems to compile ok.